contracts/core/MatchExecutor.sol

```diff
@@ -0,0 +1,369 @@
+          if (matches[i].buys.length == 1) {
+            exchange.matchOneToManyOrders(matches[i].buys[0], matches[i].sells);
+          } else if (matches[i].sells.length == 1) {
+            exchange.matchOneToManyOrders(matches[i].sells[0], matches[i].buys);
+          } else {
+            revert('invalid one to many order');
+          }
+        } else {
+          revert('invalid match type');
+        }
+        unchecked {
+          ++i;
+        }
+      }
+    }
+  }
+
+  /**
+   * @notice Function called to execute a batch of matches by calling the exchange contract
+   * @param matches The batch of matches to execute on the exchange
+   */
+  function _matchOrdersCalldata(MatchExecutorTypes.MatchOrders[] calldata matches) internal {
+    uint256 numMatches = matches.length;
+    if (numMatches > 0) {
+      for (uint256 i; i < numMatches; ) {
+        MatchExecutorTypes.MatchOrdersType matchType = matches[i].matchType;
+        if (matchType == MatchExecutorTypes.MatchOrdersType.OneToOneSpecific) {
+          exchange.matchOneToOneOrders(matches[i].buys, matches[i].sells);
+        } else if (matchType == MatchExecutorTypes.MatchOrdersType.OneToOneUnspecific) {
+          exchange.matchOrders(matches[i].buys, matches[i].sells, matches[i].constructs);
+        } else if (matchType == MatchExecutorTypes.MatchOrdersType.OneToMany) {
+          if (matches[i].buys.length == 1) {
+            exchange.matchOneToManyOrders(matches[i].buys[0], matches[i].sells);
+          } else if (matches[i].sells.length == 1) {
+            exchange.matchOneToManyOrders(matches[i].sells[0], matches[i].buys);
+          } else {
+            revert('invalid one to many order');
+          }
+        } else {
+          revert('invalid match type');
+        }
+        unchecked {
+          ++i;
+        }
+      }
+    }
+  }
+
+  /**
+   * @notice broker a trade by fulfilling orders on other exchanges and transferring nfts to the intermediary
+   * @param externalFulfillments The specification of the external calls to make and nfts to transfer
+   */
+  function _broker(MatchExecutorTypes.ExternalFulfillments memory externalFulfillments) internal {
+    uint256 numCalls = externalFulfillments.calls.length;
+    if (numCalls > 0) {
+      for (uint256 i; i < numCalls; ) {
+        _call(externalFulfillments.calls[i]);
+        unchecked {
+          ++i;
+        }
+      }
+    }
+
+    if (externalFulfillments.nftsToTransfer.length > 0) {
+      /// Transfer the nfts to the intermediary
+      _transferMultipleNFTs(address(this), intermediary, externalFulfillments.nftsToTransfer);
+    }
+  }
+
+  /**
+   * @notice broker a trade by fulfilling orders on other exchanges and transferring nfts to the intermediary
+   * @param externalFulfillments The specification of the external calls to make and nfts to transfer
+   */
+  function _brokerCalldata(MatchExecutorTypes.ExternalFulfillments calldata externalFulfillments) internal {
```

Comment:
this has same functionality as the above function

Reply:
See above